### PR TITLE
XRT-1336 branch

### PR DESCRIPTION
**Pull request by SteveOss**
Data transform functions added

### CODE REVIEW

**Review by BinduRao2018 — COMMENTED** [review]
Have left some comments inline

File: src/c/data.c [hunk 1]
```diff
@@ -625,17 +625,132 @@ bool iot_data_equal (const iot_data_t * v1, const iot_data_t * v2)
   return (iot_data_compare (v1, v2) == 0);
 }
 
+static bool iot_data_cast_val (const iot_data_union_t * in, void * out, iot_data_type_t in_type, iot_data_type_t out_type)
+{
+#define VALUE() (dfloat ? uval.f64 : (usign ? (double) uval.ui64 : (double) uval.i64))
+#define CAST(T) (*((T*) out) = dfloat ? (T) uval.f64 : (usign ? (T) uval.ui64 : (T) uval.i64))
+
+  static const double max_vals[IOT_DATA_FLOAT64] = { INT8_MAX, UINT8_MAX, INT16_MAX, UINT16_MAX, INT32_MAX, UINT32_MAX, INT64_MAX, UINT64_MAX, FLT_MAX };
+  static const double min_vals[IOT_DATA_FLOAT64] = { INT8_MIN, 0.0f, INT16_MIN, 0.0f, INT32_MIN, 0.0f, INT64_MIN, 0.0f , -FLT_MAX };
```
[the remark this answers is not in the document]
Comment (SteveOss) [reply src/c/data.c:634]: Integers have a max (positive) size and a minimum (negative) size. However floating point numbers have a maximum
precision and a minimum precision for both positive and negative numbers. Am actually checking if float number between positive maximum size and negative maximum size, really should also check if number between positive minimum size
and negative minimum size, but in this case am pretty sure will just end up mapped to zero, so sort of ok ?

File: src/c/data.c [hunk 2]
```diff
@@ -625,17 +625,132 @@ bool iot_data_equal (const iot_data_t * v1, const iot_data_t * v2)
...
+
+  bool usign = (in_type == IOT_DATA_UINT64) || (in_type == IOT_DATA_BOOL);
+  bool dfloat = (in_type <= IOT_DATA_UINT32) || (in_type == IOT_DATA_FLOAT32) || (in_type == IOT_DATA_FLOAT64);
```
[the remark this answers is not in the document]
Comment (BinduRao2018) [reply src/c/data.c:637]: ok.

**Comment by SteveOss** [comment]
recheck

**Review by BinduRao2018 — APPROVED** [review]
LGTM

**Pull request merged**